gh-457: New endpoints for copying models

mhocc opened the pull request:
Resolves #457 

Initial draft of a new endpoint for copying a data model. This is WIP, more endpoints will eventually appear, but focusing on data models for now.

# Background context

Assume that users for NHS data dictionary will want to copy an existing catalogue item (e.g. a terminology of national codes, a data set) so that they can base a new item on the original and modify, rather than create entirely from scratch. This means adding additional endpoints to support the domains.

Some other restrictions that have been agreed between @pjmonks and @jamesrwelch:

1. A model can only be copied when it is contained within a versioned folder - otherwise it should just be forked
2. A source model (and extension the parent versioned folder) cannot be finalised, it must be draft. Again, it would be better to fork otherwise.
3. A new copied model must be targeted to a sub-folder within the same parent versioned folder as the source model - it cannot be copied outside of this boundary

# Completed

- `ModelController` updated with `copyModel()` action
- URL mappings for data models currently in place to use `ModelController.copModel()`
- Checks in place to ensure that inputs are valid and preconditions are met (see above)
- Functional tests added to test everything so far
- Copying terms 
- Copying terminologies
- Copying CodeSets
- Asynchronous copying for sub-classes of DataModels

Not sure how to fully test async copying, needs some method of finding the copied item—by label?

The following need checking to see if it is applying the correct restrictions in the controller interceptors:
-   The TODO in `ModelInterceptor.checkModelActionsAuthorised()`
-   The TODO in `TermInterceptor.before()`

The copying of `DataClass` and `DataElement` types appear to have been implemented—no additional work done for these.

mhocc committed:
gh-457: Created endpoint and tests

-   The test is not currently passing and this may be because the test
    model is not valid.
-   Worked around the security restriction in `DataModelInterceptor`—needs reviewing
-   Noticed that copy permisssions is not implemented and will throw is
    requested.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MauroDataMapper/mdm-core?pullRequest=465) <br/>All committers have signed the CLA.

pjmonks committed:
gh-457 Update for copying a data model

- Found issues with initial controller action, was not committing the copy to Mauro or setting model security correctly
- Moved controller action into ModelController for other model domains to use
- Changed URL to `PUT /dataModels/{id}/copy` to look consistent with fork endpoint
- Altered the PUT request body to not need branch name or version number. This will only work within a draft versioned folder, these will be inherited
- Updated test

pjmonks committed:
gh-457 Updates to solidify copy of data models

- Ensure that new label selected is unique
- Reorganise tests
- Add additional tests to guarantee some of the edge cases - label uniqueness, source model exists, source model is not finalised

pjmonks committed:
gh-457 Guarantee preconditions of copying data models

- Must be within a draft versioned folder
- Must be copied to the same versioned folder (but can be in a different sub-folder)
- Update tests to check all conditions

jamesrwelch commented:
@pjmonks I think this looks like a sensible approach, yes 👍

mhocc committed:
gh-457: Copy terminology, terms with tests

mhocc committed:
gh-457: Add endpoint for copying CodeSet with tests

mhocc committed:
gh-457: Add async option for copying model

mhocc committed:
Merged 'develop' to resolve conflicts

pjmonks commented on mdm-plugin-terminology/grails-app/controllers/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermInterceptor.groovy:
```diff
@@ -49,6 +49,11 @@ class TermInterceptor extends TerminologySecuredInterceptor {
 
+        // TODO: is this appropriate?
+        if (actionName == 'copyTerm') {
+            return canEditModelAndReadOtherModel()
```
This isn't quite right, but the intention is correct. The URL mapping and expected request body (JSON) is this:

```
PUT /terminologies/{terminologyId}/terms/copy/{termId}

{
    "targetTerminologyId": "uuid",
    "label": "some label"
}
```

Any variables in the URL route are found in the `params` map, and anything in the request body can be found in the `request` map (I think...)

When `actionName` is `"copyTerm"`, it will need to check:

1. Can the source terminology be read? I think this is done by `currentUserSecurityPolicyManager.userCanReadSecuredResourceId(Terminology.class, params.terminologyId)`
2. Can the target terminology be edited? I think this is done by `currentUserSecurityPolicyManager.userCanEditSecuredResourceId(Terminology.class, request.targetTerminologyId)`

If either of those are `false` then return `notFound()`. It would be helpful if @joe-crawford could confirm I got this correct too.

joe-crawford replied:
@pjmonks Yes definitely agree with your logic above, and also agree that the canEditModelAndReadOtherModel() call won't work, the model IDs will be the wrong way round and `targetTerminologyId` won't be picked up.

mhocc replied:
Implemented Pete's suggestion

pjmonks commented on mdm-core/src/main/groovy/uk/ac/ox/softeng/maurodatamapper/core/interceptor/ModelInterceptor.groovy:
```diff
@@ -117,6 +117,15 @@ abstract class ModelInterceptor extends TieredAccessSecurableResourceInterceptor
 
+        // TODO: check this before merge - pure guesswork
+        if (actionName == 'copyModel') {
+            boolean canReadModel = currentUserSecurityPolicyManager.userCanReadSecuredResourceId(getSecuredClass(), getId())
```
This seems correct to me 👍

joe-crawford replied:
In ModelInterceptor, there's also this bit 

```
            // We control addition of models into containers by using container permissions
            if (isSave()) {
                return currentUserSecurityPolicyManager.userCanCreateResourceId(Model, null, Folder, params.folderId) ?:
                       forbiddenOrNotFound(canReadFolder, Folder, params.folderId)
            }
```

which checks the parent folder is writeable, should also do a similar check when copying to a new folder. Probably can tweak this to be `isSave() || isCopy()`.

mhocc replied:
Not sure if the suggested change belongs to this task.

joe-crawford replied:
Noted that it looks like there's a restriction to only copy models within their original versioned folder, however is there any check that the user can actually edit the versioned folder rather than just read it?

pjmonks reviewed (changes requested):
This is looking great to me! I've tested manually that I can copy a data model and terminology and this works. I wasn't able to test copying a term yet since the interceptor security is not letting me - please review my comment on that and fix.

The automated tests all look good too.

One question for @jamesrwelch and @joe-crawford - do these changes require a new available action to be returned for each resource in the `availableActions` array, such as `copy`? From previous experience fiddling with this that sounds complex and breaks lots of tests. Or could we make do with one of the existing ones, such as `save`?

joe-crawford commented:
> This is looking great to me! I've tested manually that I can copy a data model and terminology and this works. I wasn't able to test copying a term yet since the interceptor security is not letting me - please review my comment on that and fix.
> 
> The automated tests all look good too.
> 
> One question for @jamesrwelch and @joe-crawford - do these changes require a new available action to be returned for each resource in the `availableActions` array, such as `copy`? From previous experience fiddling with this that sounds complex and breaks lots of tests. Or could we make do with one of the existing ones, such as `save`?

Hi @pjmonks , yes aware how painful that is, definitely if we can piggy back on an existing permission/action, makes sense. Can anyone who can read the model create a copy? In that case doesn't even really need to be listed specifically?

pjmonks commented:
> > This is looking great to me! I've tested manually that I can copy a data model and terminology and this works. I wasn't able to test copying a term yet since the interceptor security is not letting me - please review my comment on that and fix.
> > The automated tests all look good too.
> > One question for @jamesrwelch and @joe-crawford - do these changes require a new available action to be returned for each resource in the `availableActions` array, such as `copy`? From previous experience fiddling with this that sounds complex and breaks lots of tests. Or could we make do with one of the existing ones, such as `save`?
> 
> Hi @pjmonks , yes aware how painful that is, definitely if we can piggy back on an existing permission/action, makes sense. Can anyone who can read the model create a copy? In that case doesn't even really need to be listed specifically?

The key part of copying is that a user is able to create a new model/item in the target location, so it's effectively the same permission as "create" (or whatever that is called). So let's ignore an additional action then.

mhocc committed:
gh-457: Fixes for PR + license header

pjmonks committed:
gh-457 Fixes for copy endpoints

- TermInterceptor security had to be adjusted to allow the copyTerm action correctly. Cannot check security of target terminology because the ID is in the request which is hard to get to, a job for another day
- Added correct tracked endpoints

pjmonks reviewed (approved):
Had to make a tweak to the `TermInterceptor` security again but my manual testing shows it works now. Ready to go ✔️

joe-crawford commented on mdm-plugin-terminology/grails-app/controllers/uk/ac/ox/softeng/maurodatamapper/terminology/item/TermInterceptor.groovy:
```diff
@@ -49,6 +52,26 @@ class TermInterceptor extends TerminologySecuredInterceptor {
+                return notFound(Terminology, params.terminologyId)
+            }
+
+            // TODO: cannot check security of target terminology because reading the request body won't allow reading the input stream a second time in
+            //  the controller. Figure out how to solve this later
+//            DataBindingSource bindingSource = cacheRequestBody()
+//            UUID targetTerminologyId = bindingSource && bindingSource.containsProperty('targetTerminologyId')
+//                ? UUID.fromString(bindingSource.getPropertyValue('targetTerminologyId').toString())
+//                : null
+//
+//            boolean canReadTarget = currentUserSecurityPolicyManager.userCanReadSecuredResourceId(Terminology, targetTerminologyId)
+//            if (!currentUserSecurityPolicyManager.userCanEditSecuredResourceId(Terminology, targetTerminologyId)) {
+//                return forbiddenOrNotFound(canReadTarget, Terminology, targetTerminologyId)
+//            }
```
If it isn't possible to get the `targetTerminologyId` out in the Interceptor, it should be possible to just add the call to `currentUserSecurityPolicyManager.userCanEditSecuredResourceId(Terminology, targetTerminologyId)` directly within the controller.

pjmonks committed:
gh-457 Double check the security access of models to copy to

pjmonks commented:
@joe-crawford I've added some additional security checks you requested, please can you check they're OK?

pjmonks commented:
Only one test failure this time, again a different test and I think due to a database error when running, so I don't think it's related to this change.

joe-crawford commented:
Great, thanks @pjmonks, @mhocc, looks good to me :+1:
